In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
users = pd.read_csv('takehome_users.csv', encoding = 'latin-1')


In [3]:
engagement = pd.read_csv('takehome_user_engagement.csv')

In [4]:
users

object_id        creation_time               name  \
0              1  2014-04-22 03:53:30     Clausen August   
1              2  2013-11-15 03:45:04      Poole Matthew   
2              3  2013-03-19 23:14:52  Bottrill Mitchell   
3              4  2013-05-21 08:09:28    Clausen Nicklas   
4              5  2013-01-17 10:14:20          Raw Grace   
...          ...                  ...                ...   
11995      11996  2013-09-06 06:14:15       Meier Sophia   
11996      11997  2013-01-10 18:28:37      Fisher Amelie   
11997      11998  2014-04-27 12:45:16        Haynes Jake   
11998      11999  2012-05-31 11:55:59       Faber Annett   
11999      12000  2014-01-26 08:57:12         Lima Thaís   

                            email     creation_source  \
0        AugustCClausen@yahoo.com        GUEST_INVITE   
1          MatthewPoole@gustr.com          ORG_INVITE   
2      MitchellBottrill@gustr.com          ORG_INVITE   
3       NicklasSClausen@yahoo.com        GUEST_INVITE   
4              GraceRaw@yahoo.com        GUEST_INVITE   
...                           ...                 ...   
11995       SophiaMeier@gustr.com          ORG_INVITE   
11996      AmelieFisher@gmail.com  SIGNUP_GOOGLE_AUTH   
11997         JakeHaynes@cuvox.de        GUEST_INVITE   
11998          mhaerzxp@iuxiw.com   PERSONAL_PROJECTS   
11999   ThaisMeloLima@hotmail.com              SIGNUP   

       last_session_creation_time  opted_in_to_mailing_list  \
0                    1.398139e+09                         1   
1                    1.396238e+09                         0   
2                    1.363735e+09                         0   
3                    1.369210e+09                         0   
4                    1.358850e+09                         0   
...                           ...                       ...   
11995                1.378448e+09                         0   
11996                1.358275e+09                         0   
11997                1.398603e+09                         1   
11998                1.338638e+09                         0   
11999                1.390727e+09                         0   

       enabled_for_marketing_drip  org_id  invited_by_user_id  
0                               0      11             10803.0  
1                               0       1               316.0  
2                               0      94              1525.0  
3                               0       1              5151.0  
4                               0     193              5240.0  
...                           ...     ...                 ...  
11995                           0      89              8263.0  
11996                           0     200                 NaN  
11997                           1      83              8074.0  
11998                           0       6                 NaN  
11999                           1       0                 NaN  

[12000 rows x 10 columns]

In [5]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   object_id                   12000 non-null  int64  
 1   creation_time               12000 non-null  object 
 2   name                        12000 non-null  object 
 3   email                       12000 non-null  object 
 4   creation_source             12000 non-null  object 
 5   last_session_creation_time  8823 non-null   float64
 6   opted_in_to_mailing_list    12000 non-null  int64  
 7   enabled_for_marketing_drip  12000 non-null  int64  
 8   org_id                      12000 non-null  int64  
 9   invited_by_user_id          6417 non-null   float64
dtypes: float64(2), int64(4), object(4)
memory usage: 937.6+ KB


In [6]:
engagement

time_stamp  user_id  visited
0       2014-04-22 03:53:30        1        1
1       2013-11-15 03:45:04        2        1
2       2013-11-29 03:45:04        2        1
3       2013-12-09 03:45:04        2        1
4       2013-12-25 03:45:04        2        1
...                     ...      ...      ...
207912  2013-09-06 06:14:15    11996        1
207913  2013-01-15 18:28:37    11997        1
207914  2014-04-27 12:45:16    11998        1
207915  2012-06-02 11:55:59    11999        1
207916  2014-01-26 08:57:12    12000        1

[207917 rows x 3 columns]

In [7]:
engagement.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207917 entries, 0 to 207916
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   time_stamp  207917 non-null  object
 1   user_id     207917 non-null  int64 
 2   visited     207917 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 4.8+ MB


In [8]:
engagement_dt = pd.to_datetime(engagement['time_stamp'])
engagement['date'] = engagement_dt.dt.date
engagement.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207917 entries, 0 to 207916
Data columns (total 4 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   time_stamp  207917 non-null  object
 1   user_id     207917 non-null  int64 
 2   visited     207917 non-null  int64 
 3   date        207917 non-null  object
dtypes: int64(2), object(2)
memory usage: 6.3+ MB


In [9]:
user_count = engagement.groupby('user_id').count()
user_3x = user_count[user_count['date'] >=3]
user_3x
engagement['user_3x'] = engagement['user_id'].apply(lambda x: 1 if x in user_3x.index else 0)

engagement


time_stamp  user_id  visited        date  user_3x
0       2014-04-22 03:53:30        1        1  2014-04-22        0
1       2013-11-15 03:45:04        2        1  2013-11-15        1
2       2013-11-29 03:45:04        2        1  2013-11-29        1
3       2013-12-09 03:45:04        2        1  2013-12-09        1
4       2013-12-25 03:45:04        2        1  2013-12-25        1
...                     ...      ...      ...         ...      ...
207912  2013-09-06 06:14:15    11996        1  2013-09-06        0
207913  2013-01-15 18:28:37    11997        1  2013-01-15        0
207914  2014-04-27 12:45:16    11998        1  2014-04-27        0
207915  2012-06-02 11:55:59    11999        1  2012-06-02        0
207916  2014-01-26 08:57:12    12000        1  2014-01-26        0

[207917 rows x 5 columns]

In [10]:
engagement_3x = engagement[engagement['user_3x'] == 1]
engagement_3x.head(20)

time_stamp  user_id  visited        date  user_3x
1   2013-11-15 03:45:04        2        1  2013-11-15        1
2   2013-11-29 03:45:04        2        1  2013-11-29        1
3   2013-12-09 03:45:04        2        1  2013-12-09        1
4   2013-12-25 03:45:04        2        1  2013-12-25        1
5   2013-12-31 03:45:04        2        1  2013-12-31        1
6   2014-01-08 03:45:04        2        1  2014-01-08        1
7   2014-02-03 03:45:04        2        1  2014-02-03        1
8   2014-02-08 03:45:04        2        1  2014-02-08        1
9   2014-02-09 03:45:04        2        1  2014-02-09        1
10  2014-02-13 03:45:04        2        1  2014-02-13        1
11  2014-02-16 03:45:04        2        1  2014-02-16        1
12  2014-03-09 03:45:04        2        1  2014-03-09        1
13  2014-03-13 03:45:04        2        1  2014-03-13        1
14  2014-03-31 03:45:04        2        1  2014-03-31        1
20  2013-01-16 22:08:03       10        1  2013-01-16        1
21  2013-01-22 22:08:03       10        1  2013-01-22        1
22  2013-01-30 22:08:03       10        1  2013-01-30        1
23  2013-02-04 22:08:03       10        1  2013-02-04        1
24  2013-02-06 22:08:03       10        1  2013-02-06        1
25  2013-02-14 22:08:03       10        1  2013-02-14        1

In [11]:
engagement_3x.groupby('user_id').count()

time_stamp  visited  date  user_3x
user_id                                    
2                14       14    14       14
10              284      284   284      284
20                7        7     7        7
33               18       18    18       18
42              342      342   342      342
...             ...      ...   ...      ...
11975           216      216   216      216
11980             8        8     8        8
11981             4        4     4        4
11988            30       30    30       30
11991             5        5     5        5

[2248 rows x 4 columns]

There are 2,248 users who have visited on 3 separate occasions.  Next, to find out how many of those users made 3 visits within a 7 day stretch.

In [12]:
engagement_3x.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201002 entries, 1 to 207909
Data columns (total 5 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   time_stamp  201002 non-null  object
 1   user_id     201002 non-null  int64 
 2   visited     201002 non-null  int64 
 3   date        201002 non-null  object
 4   user_3x     201002 non-null  int64 
dtypes: int64(3), object(2)
memory usage: 9.2+ MB


In [13]:
engagement_3x['date_num'] = pd.to_datetime(engagement_3x['date']).dt.strftime("%Y%m%d")
engagement_3x['date_num'] = pd.to_datetime(engagement_3x['date_num'], format='%Y%m%d')



<ipython-input-13-06878f0a3be7>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  engagement_3x['date_num'] = pd.to_datetime(engagement_3x['date']).dt.strftime("%Y%m%d")
<ipython-input-13-06878f0a3be7>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  engagement_3x['date_num'] = pd.to_datetime(engagement_3x['date_num'], format='%Y%m%d')


In [14]:
# engagement_3x.groupby('date')['user_id'].rolling(7).count().head(30)

engagement_3x = engagement_3x.set_index('date_num')
engagement_3x.groupby('date_num')['user_id'].resample('W').sum()


date_num    date_num  
2012-05-31  2012-06-03       5121
2012-06-01  2012-06-03       8068
2012-06-02  2012-06-03      10684
2012-06-03  2012-06-03      16606
2012-06-04  2012-06-10      23002
                           ...   
2014-05-31  2014-06-01    2636616
2014-06-01  2014-06-01    2514403
2014-06-02  2014-06-08    2410906
2014-06-03  2014-06-08    2277954
2014-06-04  2014-06-08    2200100
Name: user_id, Length: 735, dtype: int64

In [15]:
engagement_3x

time_stamp  user_id  visited        date  user_3x
date_num                                                              
2013-11-15  2013-11-15 03:45:04        2        1  2013-11-15        1
2013-11-29  2013-11-29 03:45:04        2        1  2013-11-29        1
2013-12-09  2013-12-09 03:45:04        2        1  2013-12-09        1
2013-12-25  2013-12-25 03:45:04        2        1  2013-12-25        1
2013-12-31  2013-12-31 03:45:04        2        1  2013-12-31        1
...                         ...      ...      ...         ...      ...
2014-04-20  2014-04-20 14:22:45    11991        1  2014-04-20        1
2014-04-25  2014-04-25 14:22:45    11991        1  2014-04-25        1
2014-04-28  2014-04-28 14:22:45    11991        1  2014-04-28        1
2014-05-17  2014-05-17 14:22:45    11991        1  2014-05-17        1
2014-05-30  2014-05-30 14:22:45    11991        1  2014-05-30        1

[201002 rows x 5 columns]

In [17]:
visits = engagement_3x["user_id"].value_counts()
adopted = visits[visits >= 3]
 
adopted

3623    606
906     600
1811    593
7590    590
8068    585
       ... 
7700      3
3606      3
1753      3
4187      3
3222      3
Name: user_id, Length: 2248, dtype: int64